ntoskrnl/io/iomgr/error.c

```diff
@@ -484,15 +484,35 @@ IopLogWorker(IN PVOID Parameter)
     ExFreePoolWithTag(Message, TAG_IO);
 }
 
-VOID
+/* See also ntoskrnl/io/iomgr/irp.c!IopCompleteRequest */
+static VOID
 NTAPI
-IopFreeApc(IN PKAPC Apc)
+IopAbortApc(IN PKAPC Apc)
 {
+    PAGED_CODE();
+
+    IopCompleteRequest(Apc,
+                       &Apc->NormalRoutine,
+                       &Apc->NormalContext,
+                       &Apc->SystemArgument1,
+                       &Apc->SystemArgument2);
+}
```

Comment:
Free the Apc?

Reply:
indeed!

Reply:
Hmm Do you know whether there some special treatment by IopCompleteRequest() that may require the Apc pointer to be still valid after the call?

Reply:
I think this might be a bug in Windows. They use `IopAbortRequest` here, which is the rundown routine that corresponds to IopCompleteRequest (where we don't use one). In that case, the APC doesn't need to be freed because it's part of the IRP structure. But that's not the case here.

On the other hand, Windows 10 still does this the same way so I may be missing something subtle.

Reply:
Note that when we call `KeInitializeApc()` in `IoRaiseHardError()`, we associate an Irp to it (the last parameter); so that later, if the APC is being ran down (and this callback is called), it should still be part of this IRP, and this PR adds this new `IopAbortApc()` that calls `IopCompleteRequest()` (so it sounds very similar to what you describe for windows) to complete the associated IRP (our `IopCompleteRequest` also retrieves the IRP associated to this APC).

Reply:
The last parameter is simply a context for the APC routines.
The APC itself is allocated from pool in IoRaiseHardError and nothing seems to free it in the rundown case.